src/Adapter/Product/Repository/ProductRepository.php

```diff
@@ -410,4 +465,151 @@ public function searchProducts(string $searchPhrase, LanguageId $languageId, Sho
 
         return $qb->execute()->fetchAllAssociative();
     }
+
+    /**
+     * @param ProductId $productId
+     *
+     * @return ShopId[]
+     */
+    public function getAssociatedShopIds(ProductId $productId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('id_shop')
+            ->from($this->dbPrefix . 'product_shop')
+            ->where('id_product = :productId')
+            ->setParameter('productId', $productId->getValue())
+        ;
+
+        $result = $qb->execute()->fetchAll();
+        if (empty($result)) {
+            return [];
+        }
+
+        $shops = [];
+        foreach ($result as $shop) {
+            $shops[] = new ShopId((int) $shop['id_shop']);
```

Comment:
@eternoendless on second thought since this method is not used very much for now I'm not sure it deserves to be cached The only moment when it's called is when a product is updated, it's a critical moment where we want to be sure we have up-to-date data

I'm not against adding some caching in this repository at some point, but it might be a bit early for that Since right now we only have one use case for now when we know we need to clear the cache: `CopyProductToShop` command

So I'd leave this one for later